### PR TITLE
[WIP] Remove extra bound checks for explicit range checks

### PR DESCRIPTION
**Pull request by ArtBlnd**
Currently, removing bound checks for explicit array range check on some major patterns.
such as
```
if (array.Length > 0)
{
    // Some (array) use expressions.
}
```

The explicit range check removal will be triggered if only the code scope of `if` expression is only can reach with `if` expressions, which means there is no Side-Effect on that code scope.

Related to https://github.com/dotnet/coreclr/issues/19620

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
cc @dotnet/jit-contrib

File: src/jit/rangecheck.cpp [hunk 1]
```diff
@@ -291,7 +291,59 @@ void RangeCheck::OptimizeRangeCheck(BasicBlock* block, GenTree* stmt, GenTree* t
     {
         JITDUMP("[RangeCheck::OptimizeRangeCheck] Between bounds\n");
         m_pCompiler->optRemoveRangeCheck(treeParent, stmt);
+
+        return;
+    }
+
+    /* TODO : When only enabled explicitly, also extend extra slots for handling explicit range checks.
+    * Check for previuos stmts has explicit range check. for example.
```
Comment (grant-d) [inline src/jit/rangecheck.cpp]: Nit - spelling of "previous"

Comment (ArtBlnd) [reply src/jit/rangecheck.cpp]: Oh sorry :(

**Comment by ArtBlnd** [comment]
I don't understand why using gtEffectiveVal method in `GT_COMMA` int $180` and get `GT_LCL_VAR` node without `GTF_ARRLEN_ARR_IDX `flags or something. is any way to get `GT_ARR_LENGTH` node in `GT_COMMA` node?

```
***** BB01, stmt 14
     ( 14, 12) [000010] ------------              *  STMT      void  (IL 0x002...0x006)
N009 ( 14, 12) [000009] -A-X--------              \--*  JTRUE     void
N007 (  1,  1) [000007] ------------                 |  /--*  CNS_INT   int    4 $42
N008 ( 12, 10) [000008] JA-X---N----                 \--*  LE        int    $182
N005 (  3,  2) [000159] ------------                    |  /--*  LCL_VAR   int    V03 cse0          $180
N006 ( 10,  8) [000160] -A-X--------                    \--*  COMMA     int    $180
N002 (  3,  3) [000006] ---X--------                       |  /--*  ARR_LENGTH int    $180
N001 (  1,  1) [000005] ------------                       |  |  \--*  LCL_VAR   ref    V00 arg0         u:1 $80
N004 (  7,  6) [000158] -A-X----R---                       \--*  ASG       int    $VN.Void
N003 (  3,  2) [000157] D------N----                          \--*  LCL_VAR   int    V03 cse0          $180
```

**Comment by ArtBlnd** [comment]
Also, I am renewing range check removal code based on `GT_JTRUE` so it will take a while to be fixed

**Comment by ArtBlnd** [comment]
I don't know why some array length check dump shows like this
```
***** BB01, stmt 14
N009 ( 14, 12) [000009] -A-X--------              \--*  JTRUE     void
N007 (  1,  1) [000007] ------------                 |  /--*  CNS_INT   int    4 $42
N008 ( 12, 10) [000008] JA-X---N----                 \--*  LE        int    $182
N005 (  3,  2) [000159] ------------                    |  /--*  LCL_VAR   int    V03 cse0          $180
N006 ( 10,  8) [000160] -A-X--------                    \--*  COMMA     int    $180
N002 (  3,  3) [000006] ---X--------                       |  /--*  ARR_LENGTH int    $180
N001 (  1,  1) [000005] ------------                       |  |  \--*  LCL_VAR   ref    V00 arg0         u:1 $80
N004 (  7,  6) [000158] -A-X----R---                       \--*  ASG       int    $VN.Void
N003 (  3,  2) [000157] D------N----                          \--*  LCL_VAR   int    V03 cse0 
```

It should be like
```
     (  7,  7)              [000007] ------------              *  STMT      void  (IL 0x000...0x005)
N005 (  7,  7)              [000006] ---X--------              \--*  JTRUE     void  
N003 (  1,  1)              [000004] ------------                 |  /--*  CNS_INT   int    4 $42
N004 (  5,  5)              [000005] J--X---N----                 \--*  LE        int    $182
N002 (  3,  3) CSE #01 (def)[000003] ---X--------                    \--*  ARR_LENGTH int    $180
N001 (  1,  1)              [000002] ------------                       \--*  LCL_VAR   ref    V00 arg0         u:1 $80
```

I think. whats wrong with it?

**Comment by ArtBlnd** [comment]
I'll upload tests and jit-diff result next week.
Thanks.

**Comment by ArtBlnd** [comment]
I think I have to rebase this code.
Seems have conflict with `optRemoveRangeCheck`

**Pull request closed without merging**